utility/run_cot.sh

```diff
@@ -0,0 +1,185 @@
+      ;;
+    t)
+      operation="${OPTARG}"
+      ;;
+	o)
+      if [ $(echo ${OPTARG} | egrep -c '^[0-9][0-9]*$') -eq 1 ]; then
+        osdid=${OPTARG}
+      else
+        echo
+        echo "ERROR: -o OSD must be numeric ID only."
+	error=1
+      fi
+      ;;
+    p)
+      if [ $(echo ${OPTARG} | egrep -c "^[0-9][0-9]*\.[0-9a-f][0-9a-f]*$" ) -eq 1 ]; then
+        pgid="${OPTARG}"
+      else
+        echo
+        echo "ERROR: -p parameter must be a valid Placement Group ID format (Example: 1.a7)"
+	error=1
+      fi
+      ;;
+    *)
+      echo
+      echo "Unrecognized argument: ${o}"
+      usage
+      ;;
+  esac
+done
+
+shift $((OPTIND-1))
+
+if [ $error -gt 0 ]; then
+  usage
+fi
+
+if [ -z "${osdid}" ]; then
+  echo
+  echo "ERROR: -o osdid required!"
+  usage
+fi
+
+if [ -z ${operation} ]; then
+  echo
+  echo "ERROR: -t task required!"
+  usage
+fi
+
+if [ -z "${pgid}" ]; then
+  echo
+  echo "ERROR: -p pgid required!"
+  usage
+fi
+
+echo "Paramters:"
+echo "  osdid=${osdid}"
+echo "  cephadm image=${cephadmopts}"
+echo "  pgid=${pgid}"
+echo "  operation=${operation}"
+
+
+echo "INFO: Gathering fsid"
+
+fsid=$(awk '/fsid *= */ {print $NF}' /etc/ceph/ceph.conf)
+if [ -z "${fsid}" ]; then
+  echo "ERROR: Could not retrieve cluster FSID from /etc/ceph/ceph.conf"
+  exit 1
+fi
+
+mkdir /var/log/ceph/${fsid}/osd.${osdid} &>/dev/null
+
+echo "INFO: stopping osd.${osdid}"
+systemctl disable --now ceph-${fsid}@osd.${osdid}
+cephadm unit --fsid ${fsid} --name osd.${osdid} stop
+RETVAL=$?
+if [ $RETVAL -ne 0 ]; then
+  echo "ERROR: Failed to stop osd.${osdid} - ret: $RETVAL"
+  exit $RETVAL
+fi
+
+echo "Performing ${operation} on PG: ${pgid}"
+
+if [ $operation == "pg-log-inject-dups" ]; then
+  echo "INFO: Operation selected to inject dups"
+
+  cat << EOF > /var/log/ceph/${fsid}/text.json
+[{"reqid": "client.4177.0:0", "version": "111'999999999", "user_version": "0", "generate": "100", "return_code": "0"},]
+EOF
+  echo "created test file with 100 corrupt dups to be injected"
+  cat << EOF > /var/log/ceph/${fsid}/osd.${osdid}/${operation}.${osdid}.sh
+#!/usr/bin/bash
+
+CEPH_ARGS='--no_mon_config --osd_pg_log_dups_tracked=${dups_tracked}' ceph-objectstore-tool \
+ --data-path /var/lib/ceph/osd/ceph-${osdid} --op ${operation} --pgid ${pgid} --file /var/log/ceph/text.json \
+  &> /var/log/ceph/osd.${osdid}/${operation}.${osdid}.log
+EOF
+
+else
+  cat << EOF > /var/log/ceph/${fsid}/osd.${osdid}/${operation}.${osdid}.sh
+#!/usr/bin/bash
+
+CEPH_ARGS='--osd_pg_log_trim_max=${maxtrim}' ceph-objectstore-tool --data-path /var/lib/ceph/osd/ceph-${osdid} \
+ --op ${operation} --pgid ${pgid} &> /var/log/ceph/osd.${osdid}/${operation}.${osdid}.log
+EOF
+fi
+
+chmod 755 /var/log/ceph/${fsid}/osd.${osdid}/${operation}.${osdid}.sh
+
+
+echo "INFO: Running COT operation ${operation} for osd.${osdid}"
+cephadm $cephadmopts shell --fsid ${fsid} --name osd.${osdid} /var/log/ceph/osd.${osdid}/${operation}.${osdid}.sh
+RETVAL=$?
```

Comment:
Likewise, the below check can be avoid if `set -e` is used.

Reply:
Done!